File: src/main/java/com/telegram/codex/integration/telegram/application/webhook/TelegramCommandRegistry.java
package com.telegram.codex.integration.telegram.application.webhook;

import com.telegram.codex.integration.telegram.domain.InboundMessage;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;
import java.util.regex.Pattern;

@Component
public class TelegramCommandRegistry {

    private static final List<CommandSpec> COMMANDS = List.of(
        new CommandSpec(Pattern.compile("^/start(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.START),
        new CommandSpec(Pattern.compile("^/new(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.NEW_SESSION),
        new CommandSpec(Pattern.compile("^/help(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.HELP),
        new CommandSpec(Pattern.compile("^/status(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.STATUS),
        new CommandSpec(Pattern.compile("^/session(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.SESSION),
        new CommandSpec(Pattern.compile("^/memory(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.MEMORY),
        new CommandSpec(Pattern.compile("^/forget(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.FORGET),
        new CommandSpec(Pattern.compile("^/compact(?:@[\\w_]+)?$", Pattern.UNICODE_CASE), TelegramCommand.COMPACT)
    );

    public TelegramCommandRegistry() {
    }

    public Optional<TelegramCommand> resolve(InboundMessage message) {
        return COMMANDS.stream()
            .filter(command -> command.matches(message.textOrEmpty()))
            .findFirst()
            .map(CommandSpec::command);
    }

    public enum TelegramCommand {
        START,
        NEW_SESSION,
        HELP,
        STATUS,
        SESSION,
        MEMORY,
        FORGET,
        COMPACT
    }

    private record CommandSpec(Pattern pattern, TelegramCommand command) {

        private boolean matches(String text) {
            return pattern.matcher(text).matches();
        }
    }
}

File: src/telegram/commands/telegram_command_registry.rs
use std::sync::LazyLock;

use regex::Regex;

use crate::telegram::shared::inbound_message::InboundMessage;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TelegramCommand {
    Start,
    NewSession,
    Help,
    Status,
    Session,
    Memory,
    Forget,
    Compact,
}

/// `/help` and `/help@SomeBot` both resolve to the same command.
static COMMANDS: LazyLock<Vec<(Regex, TelegramCommand)>> = LazyLock::new(|| {
    [
        ("start", TelegramCommand::Start),
        ("new", TelegramCommand::NewSession),
        ("help", TelegramCommand::Help),
        ("status", TelegramCommand::Status),
        ("session", TelegramCommand::Session),
        ("memory", TelegramCommand::Memory),
        ("forget", TelegramCommand::Forget),
        ("compact", TelegramCommand::Compact),
    ]
    .into_iter()
    .map(|(name, command)| (Regex::new(&format!(r"^/{name}(?:@[\w_]+)?$")).unwrap(), command))
    .collect()
});

#[derive(Default)]
pub struct TelegramCommandRegistry;

impl TelegramCommandRegistry {
    pub fn new() -> Self {
        Self
    }

    pub fn resolve(&self, message: &InboundMessage) -> Option<TelegramCommand> {
        let text = message.text_or_empty();
        COMMANDS.iter().find(|(pattern, _)| pattern.is_match(text)).map(|(_, command)| *command)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::telegram::shared::inbound_message::InboundMessageParams;

    fn message(text: &str) -> InboundMessage {
        InboundMessage::new(InboundMessageParams {
            chat_id: "3".to_owned(),
            message_id: 1,
            text: Some(text.to_owned()),
            user_id: "5".to_owned(),
            update_id: 1,
            ..Default::default()
        })
    }

    #[test]
    fn resolves_plain_and_mentioned_commands() {
        let registry = TelegramCommandRegistry::new();
        assert_eq!(registry.resolve(&message("/help")), Some(TelegramCommand::Help));
        assert_eq!(registry.resolve(&message("/compact@On99AppBot")), Some(TelegramCommand::Compact));
        assert_eq!(registry.resolve(&message("/new")), Some(TelegramCommand::NewSession));
    }

    #[test]
    fn ignores_non_commands() {
        let registry = TelegramCommandRegistry::new();
        assert_eq!(registry.resolve(&message("hello")), None);
        assert_eq!(registry.resolve(&message("/help me")), None);
    }
}
